R/pkg/R/functions.R

```diff
@@ -2638,3 +2638,100 @@ setMethod("sort_array",
+#' @param nbins the number of bins (optional). Default value is 10.
+#' @param df the DataFrame containing the Column to build the histogram from.
+#' @param colname the name of the column to build the histogram from.
+#' @return a data.frame with the histogram statistics, i.e., counts and centroids.
+#' @rdname histogram
+#' @family agg_funcs
+#' @export
+#' @examples 
+#' \dontrun{
+#' # Create a DataFrame from the Iris dataset
+#' irisDF <- createDataFrame(sqlContext, iris)
+#' 
+#' # Compute histogram statistics
+#' histData <- histogram(df, "colname"Sepal_Length", nbins = 12)
+#'
+#' # Once SparkR has computed the histogram statistics, the histogram can be
+#' # rendered using the ggplot2 library:
+#'
+#' require(ggplot2)
+#' plot <- ggplot(histStats, aes(x = centroids, y = counts))
+#' plot <- plot + geom_histogram(data = histStats, stat = "identity", binwidth = 100)
+#' plot <- plot + xlab("Sepal_Length") + ylab("Frequency")   
+#' } 
+setMethod("histogram",
+          signature(df = "DataFrame", col = "characterOrColumn"),
+          function(df, col, nbins = 10) {
+            # Validate nbins
+            if (nbins < 2) {
+              stop("The number of bins must be a positive integer number greater than 1.")
+            }
+
+            # Round nbins to the smallest integer
+            nbins <- floor(nbins)
+
+            # Validate col
+            if (is.null(col)) {
+              stop("col must be specified.")
+            }
+
+            colname <- col
+            x <- if (class(col) == "character") {
+              if (!colname %in% names(df)) {
+                stop("Specified colname does not belong to the given DataFrame.")
+              }
+
+              # Filter NA values in the target column
+              df <- na.omit(df[, colname])
+
+              # TODO: This will be when improved SPARK-9325 or SPARK-13436 are fixed
+              eval(parse(text = paste0("df$", colname)))
+            } else if (class(col) == "Column") {
+              # Append the given column to the dataset
+              df$x <- col
+              colname <- "x"
+              col
+            }
+
+            stats <- collect(describe(df[, colname]))
+            min <- as.numeric(stats[4, 2])
+            max <- as.numeric(stats[5, 2])
+
+            # Normalize the data
+            xnorm <- (x - min) / (max - min)
+
+            # Round the data to 4 significant digits. This is to avoid rounding issues.
+            xnorm <- cast(xnorm * 10000, "integer") / 10000.0
+
+            # Since min = 0, max = 1 (data is already normalized)
+            normBinSize <- 1 / nbins
+            binsize <- (max - min) / nbins
+            approxBins <- xnorm / normBinSize
+
+            # Adjust values that are equal to the upper bound of each bin
+            bins <- cast(approxBins -
+                         ifelse(approxBins == cast(approxBins, "integer") & x != min, 1, 0),
+                         "integer")
+
+            df$bins <- bins
```

Comment:
@shivaram @felixcheung The original DataFrame is NOT being mutated. As a matter of fact, R doesn't support passing parameters by reference, so effectively, a new copy of the DataFrame is being created every time this function is being invoked. This should, in turn, trigger the creation of a new corresponding Java object.

To illustrate this, notice the dataset doesn't change after running histogram():
```
> str(irisDF)
'DataFrame': 5 variables:
 $ Sepal_Length: num 5.1 4.9 4.7 4.6 5 5.4
 $ Sepal_Width : num 3.5 3 3.2 3.1 3.6 3.9
 $ Petal_Length: num 1.4 1.4 1.3 1.5 1.4 1.7
 $ Petal_Width : num 0.2 0.2 0.2 0.2 0.2 0.4
 $ Species     : chr "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"

> histogram(irisDF, irisDF$Sepal_Length + 1)
bins counts centroids
1     0      9      5.48
2     1     23      5.84
3     2     14      6.20
4     3     27      6.56
5     4     22      6.92
6     5     20      7.28
7     6     18      7.64
8     7      6      8.00
9     8      5      8.36
10    9      6      8.72

> str(irisDF)
'DataFrame': 5 variables:
 $ Sepal_Length: num 5.1 4.9 4.7 4.6 5 5.4
 $ Sepal_Width : num 3.5 3 3.2 3.1 3.6 3.9
 $ Petal_Length: num 1.4 1.4 1.3 1.5 1.4 1.7
 $ Petal_Width : num 0.2 0.2 0.2 0.2 0.2 0.4
 $ Species     : chr "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"
```